tests/modules/layers/test_anyprecision_optimizer.py

```diff
@@ -0,0 +1,165 @@
+        if device == "cuda":
+            model.cuda()
+
+        model_clone = deepcopy(model)
+
+        self._test_adam_equivalence(model, model_clone)
+
+    def test_bfloat16_states(self, device="cpu"):
+        """verify that AnyPrecision is running using bfloat16 for states when specified (momentum, variance)"""
+        simple_model = nn.Sequential(
+            nn.Linear(5, 10, dtype=torch.bfloat16),
+            nn.GELU(),
+            nn.Linear(10, 2, dtype=torch.bfloat16),
+        )
+        simple_model.to(torch.bfloat16)
+        simple_model.to(device)
+
+        anyprecision_adam = AnyPrecisionAdamW(
+            simple_model.parameters(),
+            variance_dtype=torch.bfloat16,
+            momentum_dtype=torch.bfloat16,
+        )
+
+        for i in range(6):
+            anyprecision_adam.zero_grad()
+            inp = torch.randn(
+                5,
+                5,
+                dtype=torch.bfloat16,
+                device=next(simple_model.parameters()).device,
+            )
+            simple_model(inp).sum().backward()
+            anyprecision_adam.step()
+
+        for group in anyprecision_adam.param_groups:
+            for p in group["params"]:
+                state = anyprecision_adam.state[p]
+                assert state["exp_avg"].dtype == torch.bfloat16
+                assert state["exp_avg_sq"].dtype == torch.bfloat16
+
+    def test_kahan_summation(self, device="cpu"):
+        """verify that AnyPrecision is properly using Kahan summation when specified (momentum, variance)"""
+        simple_model = nn.Sequential(nn.Linear(5, 10), nn.GELU(), nn.Linear(10, 2))
+        simple_model.to(torch.bfloat16)
+        simple_model.to(device)
+
+        anyprecision_adam = AnyPrecisionAdamW(
+            simple_model.parameters(),
+            variance_dtype=torch.bfloat16,
+            momentum_dtype=torch.bfloat16,
+            use_kahan_summation=True,
+            compensation_buffer_dtype=torch.bfloat16,
+        )
+
+        expected_kahan_buffer = torch.tensor(
+            [
+                [2.2888e-05, 9.5367e-04, -2.2888e-05, -9.5367e-04, 7.6294e-06],
+                [-2.2888e-05, 2.2888e-05, 9.5367e-04, -2.2888e-05, 2.2888e-05],
+                [3.0518e-05, 9.5367e-04, 3.0518e-05, 7.6294e-06, 9.5367e-04],
+                [-2.2888e-05, 2.2888e-05, -9.5367e-04, 9.5367e-04, 9.5367e-04],
+                [-9.5367e-04, 3.0518e-05, -9.5367e-04, 2.2888e-05, 0.0000e00],
+                [-3.0518e-05, -9.5367e-04, -2.2888e-05, -2.2888e-05, 2.2888e-05],
+                [2.2888e-05, 9.5367e-04, 3.0518e-05, -9.4604e-04, -2.2888e-05],
+                [-9.5367e-04, -9.5367e-04, 2.2888e-05, -9.5367e-04, -3.0518e-05],
+                [9.5367e-04, -9.5367e-04, -9.4604e-04, 9.5367e-04, 2.2888e-05],
+                [9.4604e-04, -2.2888e-05, 9.4604e-04, 9.5367e-04, 2.2888e-05],
+            ],
+            dtype=torch.bfloat16,
+            device=device,
+        )
+
+        for i in range(1):
```

Comment:
why do we need a loop if we're just iterating once?

Reply:
I've extended the test do to 2 loops (which was the original reason I had the loop) and then check the buffers and dtype afterwards. 
so the loop is relevant now. :)